Unify the univariate and multivariate TPE

HideakiImamura opened the pull request:
Depends on #2615 and #2616.

Part of works for #2614.

## Motivation
The univariate TPE is a special case of the multivariate TPE, but the implementation of them in Optuna is overwrapped now. This PR aims to resolve the redundancy.

It seems to be difficult to support full backward compatibility including the behavior when the seed is fixed. The reason is that mus, sigmas, and weights in multivariate TPE are arranged in the order of observation, while those in current univariate TPE are arranged in the order of ascending mus. In this PR, we will adapt our logic to the multivariate TPE. We will verify through benchmarking experiments that the performance of the univariate TPE is not significantly impaired by this change.

## Description of the changes
- Unify the univariate and multivariate TPE
- Fix some tests

## TODOs
- [x] Performance benchmark on kurobako

## Benchmark Results

I took a benchmark between this PR and the current master. In summary, the changes made by this PR do not significantly impair the performance of the algorithm.

### Environments:
optuna: this PR (https://github.com/optuna/optuna/pull/2618/commits/9d25958f6eee9ab804f01ae8ca2b758e0febbfdd) and the current master (https://github.com/optuna/optuna/commit/be407fdde4533c1df3fe8ec9026e884bc9d4fb15)
python: 3.8
kurobako: 0.2.9
algorithms: multivariate-tpe-master-PRUNER, multivariate-tpe-this-PR-PRUNER, tpe-master-PRUNER, tpe-this-PR-PRUNER

Each algorithm was run 100 times with the same settings, and the mean and variance of the performance were plotted.

### Results
<details>
<summary>With NopPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/119822613-d0f43c00-bf2e-11eb-80a5-4e2a4dbcffae.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/119822619-d2bdff80-bf2e-11eb-95cd-bd4fea4bf2f1.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/119822620-d3569600-bf2e-11eb-9a66-9746815b9eb4.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/119822621-d3ef2c80-bf2e-11eb-8295-a1d1f8cee4ad.png)
![nasbench-c-02fe653c3da0cb4ad55ff50834c60ca94a61fc7730a851507264ce322379ed19](https://user-images.githubusercontent.com/38826298/119822623-d487c300-bf2e-11eb-96a3-8d6d2ef14470.png)

</code>
</pre>
</details>

<details>
<summary>With MedianPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/119822654-de112b00-bf2e-11eb-97c5-365f99a72d84.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/119822660-df425800-bf2e-11eb-8ab2-b7641a3a072c.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/119822664-e0738500-bf2e-11eb-8797-03a3cdd06380.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/119822670-e10c1b80-bf2e-11eb-99bd-86ad43cbe142.png)
![nasbench-c-02fe653c3da0cb4ad55ff50834c60ca94a61fc7730a851507264ce322379ed19](https://user-images.githubusercontent.com/38826298/119822671-e1a4b200-bf2e-11eb-904a-00a407d83aa8.png)

</code>
</pre>
</details>

<details>
<summary>With HyperbandPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/119822703-e9645680-bf2e-11eb-9125-9d700e1a4b42.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/119822709-eb2e1a00-bf2e-11eb-812d-1fb40cbb7c69.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/119822713-ebc6b080-bf2e-11eb-840c-b61b247fc747.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/119822715-ec5f4700-bf2e-11eb-8a4c-eecfb5d7ac5d.png)
![nasbench-c-02fe653c3da0cb4ad55ff50834c60ca94a61fc7730a851507264ce322379ed19](https://user-images.githubusercontent.com/38826298/119822717-ecf7dd80-bf2e-11eb-8e03-73a886314670.png)

</code>
</pre>
</details>

HideakiImamura commented:
I took a benchmark between this PR and the current master. In summary, the changes made by this PR do not significantly impair the performance of the algorithm.

### Environments:
optuna: this PR (https://github.com/optuna/optuna/pull/2618/commits/9d25958f6eee9ab804f01ae8ca2b758e0febbfdd) and the current master (https://github.com/optuna/optuna/commit/be407fdde4533c1df3fe8ec9026e884bc9d4fb15)
python: 3.8
kurobako: 0.2.9

Each algorithm was run 100 times with the same settings, and the mean and variance of the performance were plotted.

### Results
<details>
<summary>With NopPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/118072559-09e7c900-b3e5-11eb-9ac8-aeb25683aa65.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/118072563-0bb18c80-b3e5-11eb-8257-41f14976cd81.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/118072565-0c4a2300-b3e5-11eb-8cdb-aded0ed371f6.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/118072568-0ce2b980-b3e5-11eb-9754-ceca798b3398.png)
![nasbench-a-f7734d47490690221086445e6217606ba300517d8fedeca9b94637ca6f49a6e9](https://user-images.githubusercontent.com/38826298/118072569-0ce2b980-b3e5-11eb-82d3-dbf954d7c46e.png)

</code>
</pre>
</details>

<details>
<summary>With MedianPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/118072601-24ba3d80-b3e5-11eb-892e-981bfca26133.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/118072603-26840100-b3e5-11eb-8eef-3e3d992f0f35.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/118072606-27b52e00-b3e5-11eb-9b42-d7fb8cdc0b5b.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/118072608-284dc480-b3e5-11eb-8c81-1d1f8168c15b.png)
![nasbench-a-f7734d47490690221086445e6217606ba300517d8fedeca9b94637ca6f49a6e9](https://user-images.githubusercontent.com/38826298/118072610-284dc480-b3e5-11eb-812b-2e09147a932e.png)

</code>
</pre>
</details>

<details>
<summary>With HyperbandPruner</summary>
<pre>
<code>

![hpo-bench-naval-ddecd9410805f2a70a5be7c2d5627c3b8b7af26160dbad87f5197a6a11d5f1b3](https://user-images.githubusercontent.com/38826298/118072625-313e9600-b3e5-11eb-8225-1acae79955bd.png)
![hpo-bench-parkinson-c54587a492ee2045a50b6b9c5db265781435c0610625a77256bad107e726d964](https://user-images.githubusercontent.com/38826298/118072631-326fc300-b3e5-11eb-9779-b9e3c6518d8f.png)
![hpo-bench-protein-71dc1cce554890df75bc2b292694102521136a26a8afd6107e56672a1f9c9e54](https://user-images.githubusercontent.com/38826298/118072633-33a0f000-b3e5-11eb-904a-5a8795261b54.png)
![hpo-bench-slice-c4296ca79fed7b86790575cb9c5b2dab8426fd1c9b028ea209def49e73d5cb70](https://user-images.githubusercontent.com/38826298/118072635-34398680-b3e5-11eb-973e-fcaf145431b0.png)
![nasbench-a-f7734d47490690221086445e6217606ba300517d8fedeca9b94637ca6f49a6e9](https://user-images.githubusercontent.com/38826298/118072637-34d21d00-b3e5-11eb-89ff-5a57c1e6ab86.png)

</code>
</pre>
</details>

HideakiImamura commented:
@himkt @Crissman Let me remove assignments of this PR since you already have a lot of assigned PRs.

@c-bata @hvy Could you review this PR if you have time?

c-bata commented:
@HideakiImamura Could you please assign another reviewer instead of me?

HideakiImamura commented:
@c-bata Sure! Let me remove the assignment.

hvy commented:
Assigned @keisuke-umezawa !

hvy commented:
Could you reduce the diff now that https://github.com/optuna/optuna/pull/2615 is merged?

hvy commented:
Sorry, could you merge the latest master again? I just merge #2616 .

HideakiImamura commented:
Thanks, @hvy. I've just merged the master branch and made this PR ready for review. PTAL.

I've noticed that I made a mistake with the benchmark. I aggregated all results of `TPESampler(multivariate=False)` and `TPESampler(multivariate=True)` in one picture. Now, I', regenerating the aggregated results.

HideakiImamura commented:
@hvy I've included the benchmark results in the PR description. Due to my mistake, I did not have the results for MedianPruner and HyperbandPruner when multivariate=True. I will run the experiment now and add it to the description later. Sorry for the confusion.

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,406 @@ class _ParzenEstimatorParameters(
+            mus[n_observations] = prior_mu
+
+            if sigmas0 is None:
+                pairs_of_observation_and_idx = np.vstack(
+                    [pairs_of_observation_and_idx, (prior_mu, n_observations + 1)]
+                )
+                pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+                    np.argsort(pairs_of_observation_and_idx[:, 0])
+                ]
+
+                pairs_of_sigma_and_idx = np.empty((n_observations + 1, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = pairs_of_sigma_and_idx[
+                    np.argsort(pairs_of_sigma_and_idx[:, 1])
+                ][:-1, 0]
+                sigmas[n_observations] = prior_sigma
+            else:
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = sigmas0 * (high - low)
+                sigmas[n_observations] = prior_sigma
         else:
-            sorted_weights = unsorted_weights[order]
-        sorted_weights /= sorted_weights.sum()
+            mus = observations
+            if sigmas0 is None:
+                pairs_of_sigma_and_idx = np.empty((n_observations, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+                sigmas = pairs_of_sigma_and_idx[np.argsort(pairs_of_sigma_and_idx[:, 1])][:, 0]
+            else:
+                sigmas = sigmas0 * (high - low)
```
📝 This `_calculate_numerical_params` is the main diff between the additions in this file and the previous removed multivariate parzen estimator.

hvy replied:
The logic here seems a bit different even for the univariate case from before. Why was it for instance that we shouldn't sort the observations before inserting the prior? Could you just briefly explain how the algorithms is kept or changes, especially for the univariate case (`sigma0` is `None`)?

HideakiImamura replied:
The algorithm for the univariate case is the same as before; in the multivariate case, the `weight` calculation and the `mus` and `sigmas` calculations are done independently. The implementation is different from before because the order of them must match the order of the `observations` passed to the init of `_ParzenEstimator`. I did the following to keep the algorithm itself unchanged. Pair the observation values with their indices, calculate the sigma using the same algorithm as before looking only at the observations, and then re-sort based on the indices.

HideakiImamura commented:
@hvy @keisuke-umezawa I completed the benchmark experiment for univariate TPE and multivariate TPE and added those results in the PR description. Please take a look.

hvy commented:
Sorry could you just resolve the conflicts caused by the constant liar PR?

hvy commented:
Are continuous parameters covered by the current benchmarks?

HideakiImamura commented:
@hvy Thanks for your review! I resolved the conflicts.

> Are continuous parameters covered by the current benchmarks?

Yes, as in [here](https://github.com/optuna/kurobako/blob/master/kurobako_problems/src/hpobench.rs#L73), kurobako's HPO bench includes not only categorical variables but also discrete and uniform variables.

hvy commented:
Hm, HPOBench contains discrete parameters but no continuous ones.

HideakiImamura commented:
Ops. `Validation MSE` is not an input but an output. HPO bench includes categorical and discrete variables. Discrete variables are internally converted to real-valued variables, so the benchmark results include the continuous case.

HideakiImamura commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -839,8 +475,8 @@ def after_trial(
 
 
 def _get_observation_pairs(
```
Thanks for the comment. As you said, we need to update the docs. 

> Missing parameters are not set to None but now skipped. I think this also changes the results of the below/above split

Oh, I didn't notice it, but you are right that the logic of split is a bit different when `group=True` and there is a conditional parameter. With this PR change, `below` will always contain gamma samples. I think this change will rather work in the direction of improving the estimation performance of the Parzen estimator for `below`.

However, it would be better to see the benchmark results when `group=True` and conditional parameters are present. In any case, since we have adapted the logic to multivariate TPE and we know that multivariate TPE achieves good performance in this situation, I think such benchmark results are good for future work.

HideakiImamura replied:
Yes, we need to update the docs.

> Missing parameters are not set to None but now skipped. I think this also changes the results of the below/above split

I didn't notice it, but you are right that the logic of split is a bit different when group=True and there is a conditional parameter. With this PR change, `below` will always contain gamma samples. I think this change will rather work in the direction of improving the estimation performance of the Parzen estimator for `below`.

However, it would be better to see the benchmark results when group=True and conditional parameters are present. In any case, since we have adapted the logic to multivariate TPE and we know that multivariate TPE achieves good performance in this situation, I think such benchmark results are good for future work.

hvy commented:
Some benchmarks with `InMemoryStorage` to compare the speed of the the default sampler `TPESampler` without any explicit arguments. It seems like this version is somewhat slower when the number of trials increase. The difference is small when using `RDBStorage` so it might be negligible but it'll likely be noticeable for `InMemoryStorage` users so I'd like some feedback from others.

<details>
<summary>`suggest_float`</summary>

### `master`

| #params | #trials | time(sec) | time/trial(ms) | time/trial/param(ms) |
| ------- | ------- | --------- | -------------- | -------------------- |
| 1 | 1 | 0.00 | 0.51 | 0.51 |
| 1 | 10 | 0.00 | 0.18 | 0.18 |
| 1 | 100 | 0.09 | 0.94 | 0.94 |
| 1 | 1000 | 2.29 | 2.29 | 2.29 |
| 2 | 1 | 0.00 | 0.36 | 0.18 |
| 2 | 10 | 0.00 | 0.24 | 0.12 |
| 2 | 100 | 0.16 | 1.56 | 0.78 |
| 2 | 1000 | 4.45 | 4.45 | 2.22 |
| 4 | 1 | 0.00 | 0.54 | 0.14 |
| 4 | 10 | 0.00 | 0.41 | 0.10 |
| 4 | 100 | 0.32 | 3.17 | 0.79 |
| 4 | 1000 | 8.80 | 8.80 | 2.20 |
| 8 | 1 | 0.00 | 0.81 | 0.10 |
| 8 | 10 | 0.01 | 0.73 | 0.09 |
| 8 | 100 | 0.63 | 6.27 | 0.78 |
| 8 | 1000 | 17.40 | 17.40 | 2.18 |
| 16 | 1 | 0.00 | 1.36 | 0.08 |
| 16 | 10 | 0.01 | 1.38 | 0.09 |
| 16 | 100 | 1.36 | 13.56 | 0.85 |
| 16 | 1000 | 36.77 | 36.77 | 2.30 |
| 32 | 1 | 0.00 | 2.44 | 0.08 |
| 32 | 10 | 0.03 | 2.68 | 0.08 |
| 32 | 100 | 2.47 | 24.72 | 0.77 |
| 32 | 1000 | 73.25 | 73.25 | 2.29 |

### PR

| #params | #trials | time(sec) | time/trial(ms) | time/trial/param(ms) |
| ------- | ------- | --------- | -------------- | -------------------- |
| 1 | 1 | 0.00 | 0.42 | 0.42 |
| 1 | 10 | 0.00 | 0.17 | 0.17 |
| 1 | 100 | 0.11 | 1.07 | 1.07 |
| 1 | 1000 | 3.28 | 3.28 | 3.28 |
| 2 | 1 | 0.00 | 0.37 | 0.19 |
| 2 | 10 | 0.00 | 0.27 | 0.14 |
| 2 | 100 | 0.21 | 2.11 | 1.06 |
| 2 | 1000 | 6.38 | 6.38 | 3.19 |
| 4 | 1 | 0.00 | 0.95 | 0.24 |
| 4 | 10 | 0.00 | 0.48 | 0.12 |
| 4 | 100 | 0.42 | 4.25 | 1.06 |
| 4 | 1000 | 12.89 | 12.89 | 3.22 |
| 8 | 1 | 0.00 | 0.79 | 0.10 |
| 8 | 10 | 0.01 | 0.75 | 0.09 |
| 8 | 100 | 0.79 | 7.88 | 0.99 |
| 8 | 1000 | 25.74 | 25.74 | 3.22 |
| 16 | 1 | 0.00 | 1.36 | 0.09 |
| 16 | 10 | 0.01 | 1.42 | 0.09 |
| 16 | 100 | 1.62 | 16.17 | 1.01 |
| 16 | 1000 | 53.64 | 53.64 | 3.35 |
| 32 | 1 | 0.00 | 2.97 | 0.09 |
| 32 | 10 | 0.03 | 2.99 | 0.09 |
| 32 | 100 | 3.24 | 32.35 | 1.01 |
| 32 | 1000 | 107.70 | 107.70 | 3.37 |

</details>

<details>
<summary>`suggest_categorical` (3 choices)</summary>

### `master`

| #params | #trials | time(sec) | time/trial(ms) | time/trial/param(ms) |
| ------- | ------- | --------- | -------------- | -------------------- |
| 1 | 1 | 0.00 | 0.41 | 0.41 |
| 1 | 10 | 0.00 | 0.16 | 0.16 |
| 1 | 100 | 0.04 | 0.41 | 0.41 |
| 1 | 1000 | 1.64 | 1.64 | 1.64 |
| 2 | 1 | 0.00 | 0.34 | 0.17 |
| 2 | 10 | 0.00 | 0.23 | 0.11 |
| 2 | 100 | 0.07 | 0.73 | 0.37 |
| 2 | 1000 | 3.15 | 3.15 | 1.57 |
| 4 | 1 | 0.00 | 0.48 | 0.12 |
| 4 | 10 | 0.00 | 0.36 | 0.09 |
| 4 | 100 | 0.14 | 1.37 | 0.34 |
| 4 | 1000 | 6.21 | 6.21 | 1.55 |
| 8 | 1 | 0.00 | 0.71 | 0.09 |
| 8 | 10 | 0.01 | 0.67 | 0.08 |
| 8 | 100 | 0.25 | 2.52 | 0.31 |
| 8 | 1000 | 12.24 | 12.24 | 1.53 |
| 16 | 1 | 0.00 | 1.22 | 0.08 |
| 16 | 10 | 0.01 | 1.24 | 0.08 |
| 16 | 100 | 0.49 | 4.92 | 0.31 |
| 16 | 1000 | 25.26 | 25.26 | 1.58 |
| 32 | 1 | 0.00 | 2.15 | 0.07 |
| 32 | 10 | 0.02 | 2.41 | 0.08 |
| 32 | 100 | 0.96 | 9.65 | 0.30 |
| 32 | 1000 | 58.92 | 58.92 | 1.84 |

### PR

| #params | #trials | time(sec) | time/trial(ms) | time/trial/param(ms) |
| ------- | ------- | --------- | -------------- | -------------------- |
| 1 | 1 | 0.00 | 0.39 | 0.39 |
| 1 | 10 | 0.00 | 0.16 | 0.16 |
| 1 | 100 | 0.06 | 0.65 | 0.65 |
| 1 | 1000 | 2.38 | 2.38 | 2.38 |
| 2 | 1 | 0.00 | 0.47 | 0.24 |
| 2 | 10 | 0.01 | 0.52 | 0.26 |
| 2 | 100 | 0.13 | 1.26 | 0.63 |
| 2 | 1000 | 5.10 | 5.10 | 2.55 |
| 4 | 1 | 0.00 | 0.46 | 0.12 |
| 4 | 10 | 0.00 | 0.40 | 0.10 |
| 4 | 100 | 0.23 | 2.29 | 0.57 |
| 4 | 1000 | 9.14 | 9.14 | 2.28 |
| 8 | 1 | 0.00 | 0.72 | 0.09 |
| 8 | 10 | 0.01 | 0.72 | 0.09 |
| 8 | 100 | 0.46 | 4.62 | 0.58 |
| 8 | 1000 | 18.23 | 18.23 | 2.28 |
| 16 | 1 | 0.00 | 1.42 | 0.09 |
| 16 | 10 | 0.01 | 1.34 | 0.08 |
| 16 | 100 | 0.89 | 8.92 | 0.56 |
| 16 | 1000 | 37.43 | 37.43 | 2.34 |
| 32 | 1 | 0.00 | 2.30 | 0.07 |
| 32 | 10 | 0.03 | 2.60 | 0.08 |
| 32 | 100 | 1.80 | 18.03 | 0.56 |
| 32 | 1000 | 77.96 | 77.96 | 2.44 |

</details>

<details>
<summary>Code</summary>

```python
import argparse
import math
import time

import sqlalchemy

import optuna


class Profile:
    def __enter__(self):
        self.start = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.end = time.time()

    def get(self):
        return self.end - self.start


def build_objective_func(n_param):
    def objective(trial):
        return sum(
            [
                #math.sin(trial.suggest_float("param-{}".format(i), 0, 1))
                int(trial.suggest_categorical("param-{}".format(i), [1, 2, 3]))
                for i in range(n_param)
            ]
        )

    return objective


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("mysql_user", type=str)
    parser.add_argument("mysql_host", type=str)
    parser.add_argument("--n-params", type=int, nargs="+", default=[1, 2, 4, 8, 16, 32])
    parser.add_argument("--n-trials", type=int, nargs="+", default=[1, 10, 100, 1000])
    args = parser.parse_args()

    storage_str = "mysql+pymysql://{}@{}/".format(args.mysql_user, args.mysql_host)

    n_params = args.n_params
    n_trials = args.n_trials

    optuna.logging.set_verbosity(optuna.logging.CRITICAL)

    print(f"| #params | #trials | time(sec) | time/trial(ms) | time/trial/param(ms) |")
    print(f"| ------- | ------- | --------- | -------------- | -------------------- |")

    for n_param in n_params:
        for n_trial in n_trials:
            engine = sqlalchemy.create_engine(storage_str)
            conn = engine.connect()
            conn.execute("commit")
            database_str = "profile_storage_t{}_p{}".format(n_trial, n_param)
            try:
                conn.execute("drop database {}".format(database_str))
            except Exception:
                pass
            conn.execute("create database {}".format(database_str))
            conn.close()

            storage = optuna.storages.get_storage(storage_str + database_str)
            sampler = optuna.samplers.TPESampler()
            # study = optuna.create_study(storage=storage, sampler=sampler)
            study = optuna.create_study(sampler=sampler)

            with Profile() as prof:
                study.optimize(
                    build_objective_func(n_param),
                    n_trials=n_trial,
                    gc_after_trial=False,
                )

            print(
                f"| {n_param} | {n_trial} | {prof.get():.2f} "
                f"| {prof.get() / n_trial * 1e3:.2f} "
                f"| {prof.get() / n_trial / n_param * 1e3:.2f} "
                "|"
            )
```

</details>

_I have not profiled the code yet._

HideakiImamura commented:
@hvy Thank you for taking the speed benchmarks. 

> Personally, as this PR should make future changes to TPE incl. performance optimizations easier, I think we can just leave this as a record and continue with the refactoring.

Totally agree. Basically, we are trying to unify the code of univariate TPE into multivariate TPE. And since the code of multivariate TPE is written regarding the code of univariate TPE, the innovations made for univariate TPE after the introduction of multivariate TPE can also be applied to multivariate TPE. So, I think it would be good to work on speedup after finishing the ongoing refactoring.

hvy commented on tests/samplers_tests/tpe_tests/test_sampler.py:
```diff
@@ -992,16 +974,19 @@ def test_constant_liar_observation_pairs(direction: str) -> None:
-    assert _tpe.sampler._get_observation_pairs(study, "x", constant_liar=False) == ([], [])
-    assert _tpe.sampler._get_observation_pairs(study, "x", constant_liar=True) == (
-        [2],
+    assert _tpe.sampler._get_observation_pairs(study, ["x"], constant_liar=False) == (
+        {"x": []},
+        [],
+    )
+    assert _tpe.sampler._get_observation_pairs(study, ["x"], constant_liar=True) == (
+        {"x": [2]},
         expected_values,
     )
```
Tests are duplicated (further down) so you can remove either.
```suggestion
```

hvy commented on tests/samplers_tests/tpe_tests/test_sampler.py:
```diff
@@ -451,7 +451,7 @@ def test_sample_independent_prior() -> None:
         assert sampler.sample_independent(study, trial, "param-a", dist) != suggestion
 
-    sampler = TPESampler(prior_weight=0.5, n_startup_trials=5, seed=0)
+    sampler = TPESampler(prior_weight=0.1, n_startup_trials=5, seed=0)
```
Is this change relevant to the change in logic? If so, how come?

HideakiImamura replied:
In this test case, the following things are happening.

First, changing the `prior_weight` will change the weights calculated in the `_ParzenEstimator`; the smaller the `prior_weight`, the smaller the weight of the point corresponding to the prior. The weights are then first used in the `sample` method, where the larger the value of weights, the more Gaussian sampling is called around the corresponding point. First, on line 448, we call `sample_independent` when `prior_weight` is the default value. In this case, with this seed value and the current PR code, Gaussian sampling is called around the point corresponding to prior. On the other hand, in line 456, we call `sample_independent` for another `prior_weight` (=0.5). With the same seed value as before and the current PR code, Gaussian sampling is still called around the point corresponding to prior in this case. Since the value of seed is shared, the results of Gaussian sampling called around the same point will be the same, and the current test case will fail because the results of independent will be equal.

There are several ways to solve this.

The first way is to create a test case where prior is not selected by setting `prior_weight` to a small value. In this way, the `sample_independent` in line 456 will not be chosen by prior, and the result of sample_independent will be different for the same seed. This is the policy that I have adopted in our current PR.

The second way is to create a test case where prior is not chosen by increasing the set of candidate points. In the current test case, we have given a total of 8 `past_trials` points. In this case, the number of points assigned to `below` is `min(np.ceil(0.1 * 8), 25) = 1`, which is only 2 points if we include the prior. These two points will be the central points to consider when calling the `sample` method of `_ParzenEstimator`, so the probability of selecting prior will be relatively high. If we simply increase the number of `past_trials` considered in the test case by a factor of 10, we get `min(np.ceil(0.1 * 80), 25) = 8`, which means that we have 9 points to consider including prior. However, I decided not to use this approach because of the risk of increasing the time required for testing.

The third approach is to delete the test case itself since the results do not vary much with a small difference in prior. However, since it is true that the algorithm behaves differently with different prior, I decided not to use it this time.

hvy commented on tests/samplers_tests/tpe_tests/test_sampler.py:
```diff
@@ -499,7 +499,7 @@ def test_sample_independent_misc_arguments() -> None:
 
     sampler = TPESampler(
-        weights=lambda i: np.asarray([i * 0.11 for i in range(7)]), n_startup_trials=5, seed=0
+        weights=lambda i: np.asarray([10 - j for j in range(i)]), n_startup_trials=5, seed=0
```
Ditto.

HideakiImamura replied:
For this test case, we actually had a problem with the previous test code. To begin with, the weight function passed to TPESampler is a function that takes a non-negative integer n and outputs an `ndarray` of length n. However, the weight passed in the previous test code outputs an `ndarray` of length 7 regardless of the given n.

It seems that in the ParzenEstimator before the refactoring, 0 was assigned to the weight even if the length of the weight was inappropriate and insufficient. ([ref](https://github.com/optuna/optuna/blob/master/optuna/samplers/_tpe/parzen_estimator.py#L113))

Since there are not that many users playing with weights, I think the impact will be limited, but should we take them into account? Should we consider them?

HideakiImamura replied:
In commit https://github.com/optuna/optuna/pull/2618/commits/1f4d6a9b1256ada98a96e9900067e12939115298, I have modified the `calculate_weights` code so that it does not destroy the code of users who use inappropriate weight functions. This commit preserves backward compatibility.

hvy commented on tests/samplers_tests/tpe_tests/test_sampler.py:
```diff
@@ -643,7 +643,7 @@ def test_sample_independent_handle_unsuccessful_states(state: optuna.trial.Trial
         study._storage.create_new_trial(study._study_id, template_trial=trial)
     trial = frozen_trial_factory(30)
-    sampler = TPESampler(n_startup_trials=5, seed=0)
+    sampler = TPESampler(n_startup_trials=5, seed=2)
```
Ditto, and also for the following seed changes.

HideakiImamura replied:
This is a similar situation to the `prior_weight` problem described earlier.

For `seed=0`, even if the states of the inserted trials are different, the center of the Gaussian sampling will be the point corresponding to the prior, and the result of the suggestion will be the same. By using different seed values, the center of the Gaussian sampling can be set to the point corresponding to the inserted trials, and the results can be reproduced differently depending on the states.

hvy replied:
Wasn't the `master` behavior kept in https://github.com/optuna/optuna/pull/2618/commits/a26b3a77bc302e4c9755a3350efabc5f01eed944 ?

HideakiImamura replied:
No, it was not. The reason is that the order of `mus`, `sigmas`, and `weights`. In the master branch, they are sorted with ascending order of `mus`, but in this PR (multivariate TPE), they are ordered as observed. It affects the sampling order in `ParzenEstimator.sample` and results in the need for a fix in test cases.

HideakiImamura replied:
Cc: @keisuke-umezawa

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
+        n_observations = self._n_observations
+        if consider_prior:
+            weights = np.zeros(n_observations + 1)
+            weights[:-1] = weights_func(n_observations)[:n_observations]
```
notes: Different implementation from `multivariate_parzen_estimator`.

hvy replied:
I think we can leave a TODO to raise if the weight function is returning an ndarray of unexpected size.

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
+        n_params = len(observations)
+
+        if n_params == 1:
+            return None
```
notes: Different implementation from multivariate_parzen_estimator.

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
-                prior_pos = 0
-                order = numpy.array([]).astype(int)
-            else:  # When mus.size is greater than 0.
-                # We decide the place of the  prior.
```
nits: double ` `.

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
-                sorted_mus = low_sorted_mus_high[1:-1]
-                sorted_mus[:prior_pos] = ordered_mus[:prior_pos]
-                sorted_mus[prior_pos] = prior_mu
-                sorted_mus[prior_pos + 1 :] = ordered_mus[prior_pos:]
```
notes: Different implementation from multivariate_parzen_estimator.

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
+            mus[:n_observations] = observations
+            mus[n_observations] = prior_mu
+
+            if sigmas0 is None:
```
notes: Different implementation from multivariate_parzen_estimator.

hvy replied:
Might aid readability to leave a comment about this line, i.e. that it's true when we're not suggesting multivariate configurations and we don't need the heuristic.

hvy replied:
We could also do something along the following to improve locality 
```
if `use_simple_sigmas`:
    if self._sigmas0 is None:
        self._sigmas0 = self._precompute_sigmas0(observations)
# And, do not call `self._precompute_sigmas0` in `__init__`.
```

keisuke-umezawa commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
-            sorted_weights = unsorted_weights[order]
-        sorted_weights /= sorted_weights.sum()
+            mus = observations
+            if sigmas0 is None:
```
notes: Different implementation from multivariate_parzen_estimator.

keisuke-umezawa reviewed (commented):
Wrote some notes to clarify differences between multivariate_parzen_estimator and parzen_estimator.

HideakiImamura commented:
@hvy @keisuke-umezawa Thanks for the mob review. I've committed https://github.com/optuna/optuna/pull/2618/commits/a26b3a77bc302e4c9755a3350efabc5f01eed944 according to the discussion. This commit includes
- Add some comments around the univariate specific codes.
- Add `multivariate` argument to `_ParzenEstimatorParameters` and use it in `_ParzenEstimator` to improve the locality.
- Keep the backward compatibility in `get_observations_pair`. If `multivariate=False`, this function returns exactly the same output as that of the master branch.
- Fix and enhance test cases, especially in the univariate case.

Please take a look.

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,415 @@ class _ParzenEstimatorParameters(
+                pairs_of_observation_and_idx = np.vstack(
+                    [pairs_of_observation_and_idx, (prior_mu, n_observations + 1)]
+                )
+                pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+                    np.argsort(pairs_of_observation_and_idx[:, 0])
+                ]
```
How about skipping the first sorting above? We're sorting twice.

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,415 @@ class _ParzenEstimatorParameters(
+        if n_observations == 0:
+            consider_prior = True
+
+        pairs_of_observation_and_idx = np.asarray(
```
I'm wondering if it's not both faster an easier to read if we create a new separate array for the indices and keep two separate ndarrays.

HideakiImamura replied:
Sounds good. I'll keep two separate ndarrays here.

hvy reviewed (changes requested):
Thanks for the updates, starting to look good. Left comments mostly on `_calculate_numerical_params`. I'll take another look at the updates tests.

Might be a good idea to create an (RFC) issue to make `_get_observation_pairs` independent of whether it's multivariate or not.

hvy commented:
https://github.com/optuna/optuna/pull/2618#issuecomment-845664050

This is a rather overhauling change to the default sampler in Optuna. Let's test continuous ones as well. We can seemingly do that with the C parameter configuration in HPOBench.

HideakiImamura commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,415 @@ class _ParzenEstimatorParameters(
+        if n_observations == 0:
+            consider_prior = True
+
+        pairs_of_observation_and_idx = np.asarray(
+            [(low, -1)] + [(x, i) for i, x in enumerate(observations)] + [(high, n_observations)]
+        )
+        pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+            np.argsort(pairs_of_observation_and_idx[:, 0])
+        ]
```
>How about conditioning these lines to only execute if not multivariate, the code will still run? A comment about why those are required might also aid readability since it's not obvious (that they are temporarily sorted to compute the sigmas but then unsorted in order to match the mus and weights). We could also add the word "sorted" in the variable name?

Sounds good. I'll follow you.

> By the way, low and high are seemingly not appended in the multivariate case. 

Yes. In the multivariate case, we ignore `observations` to compute the `sigmas`. In univariate case, we consider added `low` and `high` to compute `sigmas` for `observations` but do not compute `sigmas` for added `low` and `high`.

HideakiImamura commented:
@hvy Thanks for your careful reviews. I updated the code following your suggestions. PTAL.

I take a benchmark with the NAS bench (C) to test the floating value.

HideakiImamura commented:
@hvy @keisuke-umezawa I have benchmarked and updated the results using the latest version, modified based on the discussion during the previous mob review. There appears to be generally no difference between this PR and master. PTAL.

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,419 @@ class _ParzenEstimatorParameters(
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_numerical_params(
```
Did we forget to account for `consider_endpoints` in this method?

hvy reviewed (commented):
Thanks a lot for the updated benchmarks and plots. They look promising 👍

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,419 @@ class _ParzenEstimatorParameters(
+
+            if not multivariate:
+                assert sigmas0 is None
+                pairs_of_observation_and_idx = np.asarray(
+                    [(low, -1)]
+                    + [(x, i) for i, x in enumerate(observations)]
+                    + [(high, n_observations + 1)]
+                    + [(prior_mu, n_observations)]
+                )
+                pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+                    np.argsort(pairs_of_observation_and_idx[:, 0])
+                ]
+
+                pairs_of_sigma_and_idx = np.empty((n_observations + 1, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = pairs_of_sigma_and_idx[
+                    np.argsort(pairs_of_sigma_and_idx[:, 1])
+                ][:-1, 0]
+                sigmas[n_observations] = prior_sigma
```
Can we simplify this logic/datastructure? Some like https://github.com/optuna/optuna/commit/e2eeac76f3d3e9799ac0f432118171691b481969

HideakiImamura commented:
@hvy Thank you for your helpful and insightful reviews. I updated the code following your suggestions.
- I simplify the logic of `_calculate_numerical_params`
- I add the logic considering the `consider_endpoints` flag
- I bring up some test cases for the univariate TPE to test `consider_endpoints`, `consider_magic_clip`, and `consider_prior`.

PTAL.

HideakiImamura commented:
Note: Currently, the _ParzenEstimator calculates mus, sigmas, and weights, taking into account observations outside of [low, high] during init. The restriction to the [low, high] range is done when the sample method is called. This mechanism gives unexpected results in the estimation of sigmas.

This is due to the fact that when calculating the sigmas in _calculate_numerical_params, the mus is sorted and then the low and high are added. This can be avoided by adding low and high and then sorting mus, but changing the algorithm is out of the scope of this PR and will not be addressed here. We will leave this issue for future work.

keisuke-umezawa commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -948,9 +551,10 @@ def _get_multivariate_observation_pairs(
+        # If ``multivariate`` = True and ``group`` = True, we ignore the trials that are not
+        # included in each subspace.
+        # If ``multivariate`` = False, we consider such trials.
+        if multivariate and any([param_name not in trial.params for param_name in param_names]):
```
In my understanding, the procedure of calculation does not depend of `multivariate`, because `multivariate` is always true when `any([param_name not in trial.params for param_name in param_names])` is true. If my understanding is correct, I think we can omit `multivariate` from the arguments of this method.

HideakiImamura replied:
Sorry for the confusion. Here, we want to check `any([param_name not in trial.params for param_name in param_names])` only when `multivariate = True`. In the current master, we don't check `any([param_name not in trial.params for param_name in param_names])` when `multivariate = False`, so for the consistency, I include `multivariate` here.

HideakiImamura replied:
I updated the code comment like: `If multivariate = False, we skip the check`

keisuke-umezawa reviewed (commented):
Other than the calculation of sigmas, LGTM! I will check it again after the discussion of sigmas is settled.

HideakiImamura commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,422 @@ class _ParzenEstimatorParameters(
+        prior_mu = 0.5 * (low + high)
+        prior_sigma = 1.0 * (high - low)
+
         if consider_prior:
```
Looks great. Thanks for your insightful review!

HideakiImamura commented:
@keisuke-umezawa Thanks for your reviews! I updated the codes following your suggestions. PTAL.

hvy commented on tests/samplers_tests/tpe_tests/test_parzen_estimator.py:
```diff
@@ -204,3 +208,120 @@ def test_log_pdf_parzen_estimator(multivariate: bool) -> None:
+
+@pytest.mark.parametrize(
+    "mus, prior, magic_clip, endpoints",
+    itertools.product(
+        (np.asarray([]), np.asarray([0.4]), np.asarray([-0.4, 0.4])),  # mus
+        (True, False),  # prior
+        (True, False),  # magic_clip
+        (True, False),  # endpoints
+    ),
```
Just a comment but you can also apply `pytest.mark.parametrize` multiple times to a test function. That'll have the same product effect and might be more ideomatic.

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,404 @@ class _ParzenEstimatorParameters(
+        if consider_prior:
+            sigmas[n_observations] = prior_sigma
+
+        if not multivariate and not consider_endpoints and sorted_mus_with_endpoints.shape[0] >= 4:
+            sigmas[np.argmin(mus)] = sorted_mus_with_endpoints[2] - sorted_mus_with_endpoints[1]
+            sigmas[np.argmax(mus)] = sorted_mus_with_endpoints[-2] - sorted_mus_with_endpoints[-3]
```
Probably better (faster and less error prone) to do this right before unsorting `sigmas` as line 397 (`sigmas[:] = sorted_sigmas[np.argsort(sorted_indices)]`) since this code is executed with the default TPE configuration (`consider_endopints=False`).

hvy commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -18,114 +33,404 @@ class _ParzenEstimatorParameters(
+            sigmas[:] = sorted_sigmas[np.argsort(sorted_indices)]
 
-        # We adjust the range of the 'sigma' according to the 'consider_magic_clip' flag.
+        if consider_prior:
+            sigmas[n_observations] = prior_sigma
```
A minor but these two lines seem redundant since the sigma for the prior is again set at the bottom of this method.

hvy reviewed (commented):
Thanks for the updates. The code around the mus and sigmas are simpler now. 

By the way, are the benchmarks difficult to run? I'm wondering if we should run another round since there has been some changes.

HideakiImamura commented:
@hvy Thanks for your review! I applied all of your suggestions. PTAL.

By the way, I can re-run the benchmark experiments easily. This PR will be merged after the release of v2.8.0, so I think we should have the latest benchmark result.

hvy commented:
> By the way, I can re-run the benchmark experiments easily. This PR will be merged after the release of v2.8.0, so I think we should have the latest benchmark result.

Sounds great, let's do that. 👍

hvy commented:
Note: Benchmarks have been updated.

keisuke-umezawa reviewed (approved):
LGTM! Thank you for the long PR.

hvy reviewed (approved):
Great work and thanks, LGTM given the benchmarks and the many discussions!

hvy commented:
Updated the label to `compatibility` as per suggestion by @not522 . Thanks.
